lib/jekyll/hooks.rb

```diff
@@ -0,0 +1,103 @@
+module Jekyll
+  module Hooks
+    # Helps look up hooks from the registry by owner's class
+    OWNER_MAP = {
+      Jekyll::Site => :site,
+      Jekyll::Page => :page,
+      Jekyll::Post => :post,
+      Jekyll::Document => :document,
+    }.freeze
+
+    DEFAULT_PRIORITY = 20
+
+    # compatibility layer for octopress-hooks users
+    PRIORITY_MAP = {
+      low: 10,
+      normal: 20,
+      high: 30,
+    }.freeze
+
+    # initial empty hooks
+    @registry = {
+      :site => {
+        reset: [],
+        post_read: [],
+        pre_render: [],
+        post_write: [],
+      },
+      :page => {
+        post_init: [],
+        pre_render: [],
+        post_render: [],
+        post_write: [],
+      },
+      :post => {
+        post_init: [],
+        pre_render: [],
+        post_render: [],
+        post_write: [],
+      },
+      :document => {
+        pre_render: [],
+        post_render: [],
+        post_write: [],
+      },
+    }
+
+    # map of all hooks and their priorities
+    @hook_priority = {}
+
+    NotAvailable = Class.new(RuntimeError)
+
+    # register hook(s) to be called later
+    def self.register(owners, event, priority: nil, &block)
```

Comment:
Let me add, I'm not saying remove the `**kwd` I'm saying, lets build an API where the defaults surface.